verification/scratch/mwe1/mwe1.gobra

```diff
@@ -0,0 +1,244 @@
+    }
+
+    axiom {
+        forall r IODSpec, t Trace, s state :: {Reaches(r, t, s)} Reaches(r, t, s) ==> IsTrace(r,t)
+    }
+}
+
+// Our I/O spec. The state is the private key and the most recently received message.
+type MWE1 struct {}
+
+type state struct {
+    key int
+    lastMsg int
+}
+
+// We allow send, recv to happen at any point.
+// Declassify can only be called on a MAC tag of the most recently received message
+// generated with the private key.
+ghost
+decreases
+pure func (MWE1) Guard(s state, a Action) bool {
+    return match a {
+        // NOTE: This makess our IOD spec well-formed, as what is allowed to be
+        // declassified is now deterministic in the previous I/O actions.
+        case DeclassifyIO{?t}: t == MAC(s.key, s.lastMsg)
+        case _: true
+    }
+}
+
+ghost 
+decreases
+pure func (MWE1) Update(s state, a Action) state {
+    return match a {
+        case RecvIO{?m}: state { key: s.key, lastMsg: m }
+        case _: s
+    }
+}
+
+////////////////////////////// Trusted library I/O ////////////////////////////// 
+// cf. running-example/library/library.gobra
+
+// Receive message `m` over network.
+decreases
+requires acc(History())
+ensures acc(History()) && *History() == Snoc{old(*History()), RecvIO{m}}
+func Recv() (m int)
+
+// Send tag `t` over network.
+decreases
+requires acc(History())
+ensures acc(History()) && *History() == Snoc{old(*History()), SendIO{t}}
+func Send(t int)
+
+// Declassify tag `t`.
+ghost
+decreases
+requires acc(History())
+ensures acc(History()) && *History() == Snoc{old(*History()), DeclassifyIO{t}}
+func Declassify(t int)
+
+
+////////////////////////////// Program ////////////////////////////// 
+
+// Abstract function representing the computation of a MAC.
+// key x message -> MAC tag
+decreases
+pure func MAC(int, int) int
+
+// Receives a message, authenticates it using a MAC, and sends the resulting tag.
+// The state `s` contains the private key of this router, 
+// and the most recently received message.
+// NOTE: it should suffice here to just require Reaches(...) after the program 
+// has terminated, bc. at the moment we definitely terminate and there is no way
+// to violate the I/O spec. and "undo" this violation later on (-> safety property).
+// TODO: in the future, we should probably check this after every I/O action instead?
+// In the original example, I think this is done via the shared invariant.
+preserves acc(History()) && acc(s) && low(Pre(DefaultClassification{}, *History())) && Reaches(MWE1{}, *History(), *s)
+func authenticate(ghost s *state) {
+    m := Recv()
+    LowPost(DefaultClassification{})
+    ghost s.lastMsg = m
+
+    t := MAC(s.key, m)
+
+    /* ghost */ Declassify(t)
+    /* ghost */ LowPost(DefaultClassification{})
+    Send(t)
+    /* ghost */ LowPost(DefaultClassification{})
```

Comment:
We should have a comment here about what would happen if we, for example, forgot to call `LowPost(DefaultClassification{})` or `Declassify` in any of the three places where we currently do. The interesting point is that we would get a verification error in the postcondition instead of in the the call to Send